substrate/frame/election-provider-support/src/lib.rs

```diff
@@ -361,72 +379,87 @@ pub trait ElectionDataProvider {
 	/// Clear all voters and targets.
 	#[cfg(any(feature = "runtime-benchmarks", test))]
 	fn clear() {}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	fn set_desired_targets(_count: u32) {}
 }
 
-/// Base trait for types that can provide election
-pub trait ElectionProviderBase {
-	/// The account identifier type.
+/// An [`ElectionDataProvider`] that exposes for an external entity to request a lock/unlock on
+/// updates in the election data.
+pub trait LockableElectionDataProvider: ElectionDataProvider {
+	fn set_lock() -> data_provider::Result<()>;
+	fn unlock();
+}
+
+/// Something that can compute the result of an election and pass it back to the caller in a paged
+/// way.
+pub trait ElectionProvider {
+	/// The account ID identifier;
 	type AccountId;
 
 	/// The block number type.
 	type BlockNumber;
 
-	/// The error type that is returned by the provider.
-	type Error: Debug;
+	/// The error type returned by the provider;
+	type Error: Debug + PartialEq;
 
-	/// The upper bound on election winners that can be returned.
-	///
-	/// # WARNING
-	///
-	/// when communicating with the data provider, one must ensure that
-	/// `DataProvider::desired_targets` returns a value less than this bound. An
-	/// implementation can chose to either return an error and/or sort and
-	/// truncate the output to meet this bound.
-	type MaxWinners: Get<u32>;
+	/// The maximum number of winners per page in results returned by this election provider.
+	type MaxWinnersPerPage: Get<u32>;
+
+	/// The maximum number of backers that a single page may have in results returned by this
+	/// election provider.
+	type MaxBackersPerWinner: Get<u32>;
```

Comment:
Probably add definition of `Backers` and `Winners` in crate docs? I am pretty sure its all very confusing to people new to code. Or perhaps, we can just drop these terms and use the well known terms `Validators`  and `Nominators`?